src/logging/LazyLoggerFactory.ts

```diff
@@ -0,0 +1,41 @@
+import { LazyLogger } from './LazyLogger';
+import type { Logger } from './Logger';
+import type { LoggerFactory } from './LoggerFactory';
+
+/**
+ * Wraps over another {@link LoggerFactory} that can be set lazily.
+ * This is a singleton class, for which the instance can be retrieved using {@link LazyLoggerFactory.getInstance}.
+ *
+ * Loggers can safely be created before a {@link LoggerFactory} is set.
+ * But an error will be thrown if {@link Logger.log} is invoked before a {@link LoggerFactory} is set.
+ *
+ * This creates instances of {@link LazyLogger}.
+ */
+export class LazyLoggerFactory implements LoggerFactory {
+  private static readonly instance = new LazyLoggerFactory();
+
+  private loggerFactory: LoggerFactory | undefined;
+
+  private constructor() {
+    // Singleton instance
+  }
+
+  public static getInstance(): LazyLoggerFactory {
+    return LazyLoggerFactory.instance;
+  }
+
+  public createLogger(label: string): Logger {
+    return new LazyLogger(this, label);
+  }
+
+  public setLoggerFactory(loggerFactory: LoggerFactory | undefined): void {
+    this.loggerFactory = loggerFactory;
+  }
+
+  public getLoggerFactoryOrThrow(): LoggerFactory {
+    if (!this.loggerFactory) {
+      throw new Error('Illegal logging during initialization');
+    }
+    return this.loggerFactory;
+  }
```

Comment:
For the record: I'm not a big fan of `get` and `set` since it can make refactoring harder. But I won't complain about it further since these are used a lot in this project :-)

Reply:
Ah right! But yes, consistency indeed.

Reply:
It's not that they're used that much, except for the metadata part (and that can always be changed if necessary). But I don't really have a strong opinion on it either way.

Reply:
I seem to be running into a TS limitation when using `get`/`set`: https://github.com/microsoft/TypeScript/issues/2521

Since our setter can be undefined, but our getter can not (will throw in that case), TS gives an error because it enforces the same type.
Some solutions here:
1. Hack around this TS limitation, for example via `@ts-ignore`.
2. Also use `undefined` in getter, which will require additional type-checks elsewhere.
3. Keep using the method-form.

(My preference goes to 3)

Reply:
Why would the setter accept undefined though?

Reply:
That's also the reason why you can't set an identifier with a string in `RepresentationMetadata` unfortunately. Would definitely not do 1, am ok with 3.

Reply:
> Why would the setter accept undefined though?

If you want to reset the factory. Mainly used in unit tests at the moment.

Reply:
We could add a `reset` for that. But not strong preference. Just want to remark that `set…` is not consistent with the rest of the codebase.

Reply:
Ok, will add a `reset`.